### DISCUSSION TITLE
Welcome to IfcOpenShell Discussions!

### DISCUSSION THREAD

**Discussion by aothms** [IfcOpenShell/IfcOpenShell#1581] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by TZwielehner** [comment]
Hi,
I think it is a great idea to have a spot where all discussions can take place (concerning Ifcopenshell, of course). I happen to jump between Twitter, LinkedIn, bSi website and the osarch discussion forum (and also the github issue section).
BTW, I'm Thomas (too) :)

**Reply by aothms** [reply]
Yes it took a long time for Github to release this. Note that there is also the discussions forums on sourceforge https://sourceforge.net/p/ifcopenshell/discussion/

We should have a look to see whether these can be migrated here.

**Comment by Vertexwahn** [comment]
This prevents the creation of questions as issues. Nevertheless, for public projects like this, I prefer to create a question on StackOverflow and using the tag [`ifc-open-shell`](https://stackoverflow.com/search?q=ifcopenshell). For closed source projects this is a big win. Just my 2 cents.

**Reply by aothms** [reply]
Wow there are like over 50 issues and questions tagged on SO. I had no idea, I should have a look.

**Reply by Moult** [reply]
Wow! We need a tech support team!

**Comment by Gorgious56** [comment]
Hello, I'm a civil engineer and I am modelling places from point clouds obtained with 3D Scans using Blender. I've been recently using the BlenderBim addon to export the models in ifc format for architects (who are using Archicad, Sketchup, Revit,...). 

I love it but I'm constantly going back and forth because I keep irremediably breaking the file and connection doing mundane operations, it's a bit time consuming. :)

I am experienced in Blender API & Addons in general and I would like to contribute to softening up the edges of the Blender API interface, if there is such a need. I don't know much about the IFC in&outs (yet) but I'm slowly but surely learning :)

I'll be following the project closely, don't hesitate to ping me or point me in the right direction if you're interested in a helping hand.

Cheers !

**Reply by Moult** [reply]
Hey @Gorgious56 ! This is awesome and I would love to collaborate! The best place to collaborate closer is to find us online on https://osarch.org/chat/

I guess the first thing to do is to stop using the stable release and switch to either the daily, or ideally, straight from Git. There is a guide here: https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.6.0/src/blenderbim/docs/blenderbim/installation.rst - if you were not using the daily / git version, I suspect this will immediately give you a huge improvement in stability in the sync system.

Getting native authoring in Blender is something I really want to solve. For example, only recently did we implement "undo". Which meant that if you used "undo" with the last stable release, it was likely to break the file. Crazy, I know. Because we do everything with Native IFC, there is a delicate system of synchronisation between Blender and the Native IFC systems ... over time, as I spot issues, make this sync more robust, but I need a lot more testing, and ideally testing from someone with your skillset who will also pay attention to weird syncing things. There is a lot I have learned about the quirks of Blender's undo system as well as IfcOpenShell's handling of data.

Would love to chat further, and can't wait to fix the bugs you unearth from your use of the BlenderBIM Add-on.

Also of interest: https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.6.0/src/blenderbim/docs/blenderbim/developer_guide.rst

... and perhaps if you like to read more about "undo": https://raw.githubusercontent.com/IfcOpenShell/IfcOpenShell/v0.6.0/src/blenderbim/docs/blenderbim/undo_system.rst

**Reply by Moult** [reply]
By the way, I am only awake during Sydney timezone so if you post in the chat and I don't respond, it's probably because i'm AFK :) You might want to try Matrix to help stay online and see past responses.

**Comment by Borna07** [comment]
Hello! I am looking for a way to extract general informations like when the modell is created, who created it and the name of the project form an ifc file using python. I stumbeld on a class IfcCobieParser class that contains a "get_contacts" method and it looks like it covers what I need, except for project name but I allready got that covered. I cant seem to make it work. Is it intended for the same purpouse I am trying to use it? Could you show me an code example of how I should utilize this code? I am still a Python beginner

**Reply by Moult** [reply]
I highly recommend this article: https://wiki.osarch.org/index.php?title=IfcOpenShell_code_examples

It'll take you through the basics of writing your own Python scripts. To select the project, you can do:

```python
ifc = ifcopenshell.open('/path/to/your.ifc')
project = ifc.by_type('IfcProject')[0] # Get the first IfcProject in your IFC data
print(project.get_info()) # print all data
print(project.Name) # print project name
```

Some of your information may be in the IFC header instead. For example:

```python
print(ifc.wrapped_data.header.file_name.time_stamp)
```

**Reply by Borna07** [reply]
Thanks for a quick response :) I got everything what I need

**Comment by onceadrog** [comment]
I'd be happy to tackle tooltips, as I figure things out myself.

I've made some local changes to get some working, and understand the structure.

I also noted that there are many comments, including in deprecated components (like dumb wall modules) which would be helpful for user documentation (which I'd also be happy to do work on).

I'll probably write something locally next week to scrape and categorise comments and locate missing tooltips (i.e. all of them :) )

Maybe people could put placeholder tooltips as they're coding with "TODO" so they are easy to locate and do? e.g.
``` 
class SplitWall(bpy.types.Operator):
    bl_idname = "bim.split_wall"
    bl_label = "Split Wall"
    bl_description = "Split wall to do what? TODO"
    bl_options = {"REGISTER", "UNDO"}
```

**Reply by Moult** [reply]
A small Python script that looks for the case sensitive word "Operator" and checks if bl_description appears before the next occurrence of "def" would be a good non intrusive way of creating a todo list without creating TODOs everywhere :) If you would be so kind to write a script that does this and produce a list, that would be amazing!

**Reply by onceadrog** [reply]
I've created a gist for it. It doesn't check if there is a bl_description, but grabs the rest, including all comments whether done with # or """ blocks. [Link to Comment Crawler gist](https://gist.github.com/onceadrog/d6f93aa63a553a6d6ba1c93add973e88)

I might tweak it later in the week to just grab things needing tooltips later in the week.

**Comment by onceadrog** [comment]
That's the plan, make a list based on the previously identified class, and
also list comments within each class.

**Comment by atmatth** [comment]
Hello, I am interested in the infrastructure side of ifc with focus on tunnelling.
As a start I would like to contribute to the alignment stuff. A while ago I implemented some alignment curves as 3D-bezier-curves in freecad as a macro, maybe that is of interest for ifcopenshell too? 

At the moment I am a little bit lost in all the classes and how they are connected together. Maybe someone has a hint where to start with alignment-related stuff...

Regards,
Matthias

**Reply by Vertexwahn** [reply]
Example files and documentation can be found for instance [here](https://standards.buildingsmart.org/IFC/RELEASE/IFC4_1/FINAL/HTML/link/horizontal-alignment.htm). Have also a look on [this  paper](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.702.823&rep=rep1&type=pdf).

**Reply by aothms** [reply]
Hi @atmatth there is a little bit of work already done on this front by myself and @civilx64 https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.6.0/src/ifcopenshell-python/ifcopenshell/alignment.py and https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.6.0/src/ifcopenshell-python/ifcopenshell/transition_curve.py Currently this uses a piecewise linear approach. It'd be amazing if you are able to extend this covering more of the schema or using supplementary approaches. Since the FreeCAD BIM workbench also uses IfcOpenShell it'd also benefit FreeCAD. For discussion see https://github.com/IfcOpenShell/IfcOpenShell/issues/842 and https://github.com/IfcOpenShell/IfcOpenShell/pull/1480

**Reply by Vertexwahn** [reply]
You can find also coe for the computation of different alignment segments [here](https://github.com/tumcms/Open-Infra-Platform/tree/8bf50e878c7613280f4174886a476af4b4e5d019/CurrentlyExcluded/Infrastructure/Alignment/HorizontalAlignment)

**Comment by RickBrice** [comment]
Hello, I develop suite of open source bridge engineering applications called [BridgeLink](https://github.com/WSDOT/BridgeLink). The main product in the suite is called [PGSuper](https://github.com/WSDOT/PGSuper) for **P**recast, prestressed concrete **G**irder bridge **Super**structure designs.

My interest is to add seamless IFC Bridge import and export capabilities to PGSuper. IfcOpenShell has given me huge head start in this effort.

**Comment by Mannshoch** [comment]
what could be the cause, that my Convert to STEP never ending? I did several tests but it end never, every time: https://github.com/IfcOpenShell/IfcOpenShell/discussions/2184#discussioncomment-2686162

**Reply by Moult** [reply]
I've seen this happen with either:

 - Crazy booleans (I found a model with an object with something like 8,000 booleans)
 - Crazy breps / tesselations

Can you provide the file? @aothms has a private email at models@ifcopenshell.org if that helps.

**Reply by Mannshoch** [reply]
Sorry but I work with Customer Data here. So I'm not allowed to transfer them to other.

May try to test the File  _Clinic_Plumbing_  from The [Building Smart Sample Files](https://github.com/buildingSMART/Sample-Test-Files)
this File is way smaller but I get the same effect.

**Reply by Mannshoch** [reply]
@Moult Any updates here?

**Reply by Moult** [reply]
@Mannshoch Thanks for the reminder! I tested loading the Clinic_Plumbing model in Blender and it loads in 12 seconds. Doing `IfcConvert -j 12 Clinic_Plumbing.ifc foo.glb` also takes 12 seconds.

In contrast, doing `IfcConvert -j 12 Clinic_Plumbing.ifc foo.obj` took 1m40s and `IfcConvert -j 12 Clinic_Plumbing.ifc foo.stp` took much longer that I stopped waiting. The CPU usage climbed down to 1 core though, suggesting that there are a limited set of particular objects that cause it to be upset.

Unfortunately this is a question for @aothms. My assumption is that mesh-based geometry is easy and fast in contrast to trying to intelligently transfer parametric geometry.

**Reply by Mannshoch** [reply]
Thanks for the -j 12 i was not aware I should setup count of CPU threads. I assumed it automatically use max possible. that way Its way faster.

If I use it on Clinic_Plumbing.ifc it calculate really fast to the below shown state. but it never finish and create the step file.
```
\Clinic_Plumbing.ifc test2.stp
IfcOpenShell IfcConvert v0.7.0-e508fb44 (OCC 7.5.3)
Scanning file...
Done scanning file
Parsing input file took 5 seconds
Creating geometry...
Done creating geometry (6587 objects)
[##################################################]
```

**Reply by aothms** [reply]
We rely on the OCCT STP exporter and indeed it generally doesn't cope well with complete building models. People sometimes use the STP writer for a single beam to automate manufacturing for example. Maybe setting `write.surfacecurve.mode` [0] to off would help with performance.

[0] https://dev.opencascade.org/doc/overview/html/occt_user_guides__step.html#occt_step_3_3_2

**Comment by Mannshoch** [comment]
I see that ifcconvert create STP version AP214 is there a chance to support AP242 also?

**Reply by aothms** [reply]
According to the OCCT docs [0] we can choose


1. or AP214CD (default): AP214, CD version (dated 26 November 1996),
2. or AP214DIS: AP214, DIS version (dated 15 September 1998).
3. or AP203: AP203, possibly with modular extensions (depending on data written to a file).
4. or AP214IS: AP214, IS version (dated 2002)
5. or AP242DIS: AP242, DIS version.


https://dev.opencascade.org/doc/overview/html/occt_user_guides__step.html#occt_step_3_3_2

**Reply by aothms** [reply]
I'll try to find some time to expose some more control using the command line.

**Reply by Mannshoch** [reply]
Sounds nice, may something like  `.\IfcConvert.exe STEP [commands]`

so you can `.\IfcConvert.exe STEP --help` and only get STEP related commands.
May some STEP related commands from Opencascade are available.

**Reply by aothms** [reply]
That would require dramatic changes the program options usage in IfcConvert. Not disagreeing, we have a lot of similar situations with arguments specific to a certain output format. But I don't think it's going to happen any time soon.

**Comment by Babaiqa** [comment]
Hello everyone!
I don't know much about ifc, but I have to write here, I don't think the question requires an issue. When converting large ifc files to glb, IfcConvert consumes a lot of memory. I want to optimize the converter and include in the conversion only those elements that are necessary for forming the geometry. Is there any useful information where I can find out about this?
I would be very grateful!

**Reply by aothms** [reply]
> include in the conversion only those elements that are necessary for forming the geometry

You can do something like `--include entities IfcWall` to only include walls. You can also filter based on attribute values, layers, etc. But that really depends on your use case.

We're also working on reducing memory consumption in 0.8. Mostly on the parsing side. And there is a community project ton a more light weight geometry library.

**Reply by Babaiqa** [reply]
Probably my understanding of possible inclusion of elements was a bit wrong. I would exclude elements that do not participate in geometry via "exclude", but this is just an analogue of your suggestion, and I cannot have information about the contents of files in the future.
If I understood correctly, then the only solution with preserving "quality" and completeness of data after conversion is to wait for optimization in the next version or use some other more lightweight library that you mentioned?

**Reply by Moult** [reply]
IfcConvert already only processes elements that have geometry, if that's what you mean. It won't process elements that have no geometry.

**Reply by Mannshoch** [reply]
I assume here he means only not tessellated geometry. 

In my former job I had the problem that customer got IFC files with whole buildings and any equipment down to some staged plants. But the customers only needed the some pipes or a specific room or the steel beam of a roof and were not able to extract them from the IFC. (most of my customers got IFC the first time and never worked with them). The creator of the IFC mostly does not know enough about IFC themselves. But I detected, that in some files there is a difference. You get the whole building as tessellated data, but the requested parts are geometrical data. 
Until I left I had to propose some third party apps the customers have to buy to get these data exported. I was not able to do it with IfcOpenshell or FreeCAD.

**Comment by Tyzze86** [comment]
Hi, 

I work with railways and have a former background as a formwork carpenter and a railway alignment engineer before I switched to BIM. I work mostly with information management and standardization. 

I will help updating documentation where I can (first PR coming soon). 

I would also like to contribute to make it easier to make information requirments and I have some ideas. As someone who makes a lot of information requirments I (and my colleages) miss a desktop app for this. We also see that almost all free web apps only can work with one file at a time. I will post more about this in the ideas section of the discussion. I have an idea of what I want such tool to do and also a way to implement it in stages.

**Reply by aothms** [reply]
Welcome :)

> miss a desktop app for this

Stay tuned! End of this month we're planning to release a first preview of our new desktop viewer. Over time (but not initially) it will integrate with our modules for IDS validation and CSV data extraction/integration.

**Reply by Tyzze86** [reply]
@aothms: Vailidation is not where we see the biggest challenge today. We have good experience running validation in Solibri (but the latest price increase when they are now sending office to legacy has us lookin other places) and OpenIFCViewer. 

For me it's about the creation of the requirments in for a single use case or all use cases of a project and then export these out to IDS or even according to ISO 7817-series Level of Information Need (In part two this specifically mention that IDS can be a part of these requirements for alphanummerical information)

We are looking forward to see what the new viewer can do for us and if it can actually replace Solibri in its entirity.

**Reply by Mannshoch** [reply]
> Welcome :)
> 
> > miss a desktop app for this
> 
> Stay tuned! End of this month we're planning to release a first preview of our new desktop viewer. Over time (but not initially) it will integrate with our modules for IDS validation and CSV data extraction/integration.

Near the same time as you answered here https://github.com/fougue/mayo/ released the next version after a long time.

Are you may working together with them?
